[RLlib] Checkpoint and restore connectors.

gjoliver opened the pull request:
Plus a couple of examples showing the usage of connector enabled policies.

## Why are these changes needed?

Allow checkpoint and restore of connector pipelines.

## Related issue number

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

sven1977 commented on rllib/connectors/connector.py:
```diff
@@ -365,6 +377,11 @@ def prepend(self, connector: Connector):
         """
         self.connectors.insert(0, connector)
 
+        print(
```
same

gjoliver replied:
done

sven1977 commented on rllib/connectors/connector.py:
```diff
@@ -373,6 +390,11 @@ def append(self, connector: Connector):
         """
         self.connectors.append(connector)
 
+        print(
```
same

gjoliver replied:
done

sven1977 commented on rllib/examples/connectors/adapt_connector_policy.py:
```diff
@@ -0,0 +1,118 @@
+"""This example script shows how to load a connector enabled policy,
+and adapt/use it with a different version of environment.
```
nit: "the environment"

gjoliver replied:
done

sven1977 commented on rllib/policy/policy.py:
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
             # The current global timestep.
             "global_timestep": self.global_timestep,
         }
+        if self.config.get("enable_connectors", False):
```
Can you add a one-line comment here? that we are adding the connector state?

gjoliver replied:
sure

sven1977 commented on rllib/policy/policy.py:
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
+            state["connector_configs"] = connector_configs
         return state
 
+    @ExperimentalAPI
```
Should we use `PublicApi(alpha)` here as well?

gjoliver replied:
👍

sven1977 commented on rllib/policy/policy.py:
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
+            state: The new state to set this policy to. Can be
+                obtained by calling `self.get_state()`.
+        """
+        # To avoid a circular dependency problem cause by SampleBatch.
```
Future note to ourselves: We should move SampleBatch out of the policy folder. It doesn't belong there.

gjoliver replied:
yeah, totally, SampleBatch should be a pretty low level util that depends on nothing.

sven1977 commented on rllib/policy/policy.py:
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
+            self.agent_connectors = restore_connectors_for_policy(
+                self, connector_configs["agent"]
+            )
+            print("restoring agent connectors:")
```
logger.info

gjoliver replied:
done

sven1977 commented on rllib/policy/policy_map.py:
```diff
@@ -118,48 +118,18 @@ def create_policy(
-        else:
-            class_ = policy_cls
-            self[policy_id] = class_(observation_space, action_space, merged_config)
+        _class = get_tf_eager_cls_if_necessary(policy_cls, merged_config)
```
nice! thanks for cleaning this up and creating the utility function

gjoliver replied:
👌

sven1977 reviewed (approved):
Awesome PR @gjoliver , thanks for the examples on connectors.
Just a few nits, then we can merge.

Oh, one last thing, we should add the new examples to BUILD!

gjoliver commented on rllib/examples/connectors/run_connector_policy.py:
```diff
@@ -0,0 +1,54 @@
+"""This example script shows how to load a connector enabled policy,
+and adapt and use it with a different version of environment.
```
updated the comment.
thanks.

gjoliver commented on rllib/policy/policy.py:
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
+                self, connector_configs["action"]
+            )
+            print("restoring action connectors:")
+            print(self.action_connectors.__str__(indentation=4))
```
done

gjoliver commented on rllib/connectors/connector.py:
```diff
@@ -357,6 +364,11 @@ def insert_after(self, name: str, connector: Connector):
             raise ValueError(f"Can not find connector {name}")
         self.connectors.insert(idx + 1, connector)
 
+        print(
```
done

gjoliver commented on rllib/connectors/connector.py:
```diff
@@ -341,6 +343,11 @@ def insert_before(self, name: str, connector: Connector):
             raise ValueError(f"Can not find connector {name}")
         self.connectors.insert(idx, connector)
 
+        print(
```
done

gjoliver commented:
ok, addressed all the comments. let's see if CI is happy.
completely understand your comments about logger.info(), was a bit hesitant because you often can't see log files on disks on Anyscale :).

Also added all the examples as unit tests.
I also added a multi-agent example, where we restore a TF PPO policy to train a new Torch SAC policy.
